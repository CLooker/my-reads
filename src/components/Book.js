import React from 'react';

const convertToHttps = str =>
  str.includes('https') ? str : `${str.slice(0, 4)}s${str.slice(4)}`;

const checkForImageLinks = ({ imageLinks }) => (imageLinks ? true : false);

const returnImageAddress = book =>
  checkForImageLinks(book)
    ? convertToHttps(book.imageLinks.thumbnail)
    : encodeURI(
        `https://via.placeholder.com/128x193?text=${book.title.toUpperCase()}`
      );

const returnSelectValue = shelf => (shelf ? true : ' ');

const Book = ({ book, changeBookshelf, displaySearch }) => {
  const { title, id, shelf, authors } = book;
  return (
    <div>
      <div className="book">
        <div className="book-top">
          <div
            className="book-cover"
            style={{
              backgroundImage: `url(${returnImageAddress(book)})`
            }}
          />
          <div className="book-shelf-changer">
            <select
              onChange={changeBookshelf}
              value={returnSelectValue(shelf)}
              data={id}
            >
              <option value="none" disabled>
                Move to...
              </option>
              {displaySearch && <option value=" " disabled />}
              <option value="currentlyReading">Currently Reading</option>
              <option value="wantToRead">Want to Read</option>
              <option value="read">Read</option>
              {shelf && <option value="none">None</option>}
            </select>
          </div>
        </div>
        <div className="book-title">{title}</div>
        <div className="book-authors">
          {authors &&
            authors.map(author => (
              <div key={`${author}${id}`.replace(/\s/g, '')}>{author}</div>
            ))}
        </div>
      </div>
    </div>
  );
};

export default Book;
