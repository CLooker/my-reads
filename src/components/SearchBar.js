import React from 'react';
import { Link } from 'react-router-dom';
import { DebounceInput } from 'react-debounce-input';
import PropTypes from 'prop-types';

const SearchBar = ({ query, handleChange }) => (
  <div className="search-books">
    <div className="search-books-bar">
      <Link className="close-search" to="/my-reads">
        Close
      </Link>
      <div className="search-books-input-wrapper">
        <DebounceInput
          minLength={1}
          debounceTimeout={300}
          type="text"
          placeholder="Search by title or author"
          value={query}
          onChange={handleChange}
        />
      </div>
    </div>
    <div className="search-books-results">
      <ol className="books-grid" />
    </div>
  </div>
);

SearchBar.propTypes = {
  query: PropTypes.string,
  handleChange: PropTypes.func.isRequired
};

export default SearchBar;
