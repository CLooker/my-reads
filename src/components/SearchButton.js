import React from 'react';
import { Link } from 'react-router-dom';

const SearchButton = ({ resetSearch, ...location }) => (
  <div className="open-search">
    <Link onClick={resetSearch} to="my-reads/search">
      Add a book
    </Link>
  </div>
);

export default SearchButton;
