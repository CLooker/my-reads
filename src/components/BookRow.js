import React from 'react';
import Book from './Book';
import PropTypes from 'prop-types';

const BookRow = ({
  myReads,
  shelfTitle,
  shelf,
  changeBookshelf,
  displaySearch
}) => (
  <div className="list-books">
    <div className="list-books-title">
      <h1>{myReads}</h1>
    </div>
    <div className="list-books-content">
      <div className="bookshelf">
        <h2 className="bookshelf-title">{shelfTitle}</h2>
        <div className="bookshelf-books">
          <ol className="books-grid">
            {shelf.map(book => (
              <li key={book.id}>
                <Book
                  book={book}
                  changeBookshelf={changeBookshelf}
                  displaySearch={displaySearch}
                />
              </li>
            ))}
          </ol>
        </div>
      </div>
    </div>
  </div>
);

BookRow.propTypes = {
  myReads: PropTypes.string,
  shelfTitle: PropTypes.string.isRequired,
  shelf: PropTypes.array.isRequired,
  changeBookshelf: PropTypes.func.isRequired,
  displaySearch: PropTypes.bool
};

export default BookRow;
