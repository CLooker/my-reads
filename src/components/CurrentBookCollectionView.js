import React from 'react';
import BookRow from './BookRow';
import SearchButton from './SearchButton';
import PropTypes from 'prop-types';

const CurrentBookCollectionView = ({ returnMyReadsProps, ...routeProps }) => (
  <div>
    <BookRow {...returnMyReadsProps('Currently Reading', routeProps)} />
    <BookRow {...returnMyReadsProps('Want To Read', routeProps)} />
    <BookRow {...returnMyReadsProps('Read', routeProps)} />
    <SearchButton {...routeProps} />
  </div>
);

CurrentBookCollectionView.propTypes = {
  returnMyReadsProps: PropTypes.func.isRequired
};

export default CurrentBookCollectionView;
