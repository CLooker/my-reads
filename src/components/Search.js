import React, { Component } from 'react';
import SearchBar from './SearchBar';
import BookRow from './BookRow';

class Search extends Component {
  render() {
    const {
      closeSearch,
      search,
      handleChange,
      query,
      searchResults,
      changeBookshelf
    } = this.props;
    return (
      <div>
        <SearchBar
          closeSearch={closeSearch}
          search={search}
          handleChange={handleChange}
          query={query}
        />
        <BookRow
          myReads=""
          shelfTitle="Search Results"
          shelf={searchResults}
          changeBookshelf={changeBookshelf}
          displaySearch={true}
        />
      </div>
    );
  }
}

export default Search;
