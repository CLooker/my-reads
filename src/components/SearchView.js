import React, { Component } from 'react';
import SearchBar from './SearchBar';
import BookRow from './BookRow';
import * as BooksAPI from './BooksAPI';
import PropTypes from 'prop-types';

export default class SearchView extends Component {
  static propTypes = {
    shelfKeyValueStore: PropTypes.object.isRequired,
    setStateOnApiReturn: PropTypes.func.isRequired,
    getShelfAndRender: PropTypes.func.isRequired
  };

  static getDerivedStateFromProps(nextProps, prevState) {
    let flag = false;
    const newSearchResults = Object.keys(nextProps.shelfKeyValueStore)
      .map(key => nextProps.shelfKeyValueStore[key])
      .map(book => {
        const target = book.id;
        const match = prevState.searchResults.find(({ id }) => id === target);
        if (match) {
          if (match.shelf !== target.shelf) {
            flag = true;
          }
        }
        return book;
      });

    if (flag === true) {
      return {
        query: prevState.query,
        searchResults: prevState.searchResults.map(book => {
          const target = book.id;
          const newVersion = newSearchResults.find(({ id }) => id === target);
          if (newVersion) {
            return newVersion;
          } else {
            return book;
          }
        })
      };
    } else {
      return null;
    }
  }

  state = {
    query: '',
    searchResults: []
  };

  resetSearch = () => this.setState({ searchResults: [] });

  search = query =>
    query
      ? BooksAPI.search(query).then(this.handleSearchReturn)
      : this.resetSearch();

  handleSearchReturn = res =>
    console.log('search returned this: ', res) ||
    (res !== undefined
      ? this.setState({
          searchResults: res.map(
            bookObj => this.props.shelfKeyValueStore[bookObj.id] || bookObj
          )
        })
      : this.resetSearch());

  changeBookshelf = e =>
    BooksAPI.update(
      { id: e.currentTarget.getAttribute('data') },
      e.target.value.trim()
    ).then(this.props.setStateOnApiReturn);

  handleChange = e =>
    this.setState({ query: e.target.value }, () =>
      this.search(this.state.query)
    );

  returnSearchBarProps = () => ({
    handleChange: this.handleChange,
    query: this.state.query
  });

  returnBookRowProps = () => ({
    shelfTitle: 'Search Results',
    shelf: this.state.searchResults,
    changeBookshelf: this.changeBookshelf,
    displaySearch: true
  });

  render() {
    return (
      <div>
        <SearchBar {...this.returnSearchBarProps()} />
        <BookRow {...this.returnBookRowProps()} />
      </div>
    );
  }
}
