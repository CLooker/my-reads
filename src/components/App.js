import React, { Component } from 'react';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import * as BooksAPI from './BooksAPI';
import SearchView from './SearchView';
import CurrentBookCollectionView from './CurrentBookCollectionView';
import NoMatch from './NoMatch';
import '../App.css';

export default class App extends Component {
  state = {
    shelf: [],
    shelfKeyValueStore: {},
    currentlyReading: [],
    wantToRead: [],
    read: []
  };

  componentDidMount() {
    this.getShelfAndRender();
  }

  getShelfAndRender() {
    BooksAPI.getAll().then(this.setStateOnApiReturn);
  }

  setStateOnApiReturn = apiReturn =>
    Array.isArray(apiReturn)
      ? this.buildState(apiReturn)
      : this.getShelfAndRender();

  buildState = apiReturn =>
    this.setState({
      shelf: apiReturn,
      shelfKeyValueStore: this.buildKVS(apiReturn),
      currentlyReading: this.buildShelfTypes(apiReturn, 'currentlyReading'),
      wantToRead: this.buildShelfTypes(apiReturn, 'wantToRead'),
      read: this.buildShelfTypes(apiReturn, 'read')
    });

  buildKVS = apiReturn =>
    apiReturn.reduce(
      (keyValueStore, book) => ({
        ...keyValueStore,
        [book.id]: book
      }),
      {}
    );

  buildShelfTypes = (apiReturn, type) =>
    apiReturn.filter(({ shelf }) => shelf === type);

  changeBookshelf = e =>
    console.log('changeBookshelf called') ||
    BooksAPI.update(
      { id: e.currentTarget.getAttribute('data') },
      e.target.value.trim()
    ).then(this.setStateOnApiReturn);

  returnSearchViewProps = routeProps => ({
    shelfKeyValueStore: this.state.shelfKeyValueStore,
    setStateOnApiReturn: this.setStateOnApiReturn,
    getShelfAndRender: this.getShelfAndRender,
    ...routeProps
  });

  returnMyReadsProps = (shelfTitle, routeProps) =>
    shelfTitle === 'Currently Reading'
      ? {
          myReads: 'MyReads',
          shelfTitle,
          shelf: this.state.currentlyReading,
          changeBookshelf: this.changeBookshelf,
          selectedValue: 'currentlyReading',
          ...routeProps
        }
      : shelfTitle === 'Want To Read'
        ? {
            shelfTitle,
            shelf: this.state.wantToRead,
            changeBookshelf: this.changeBookshelf,
            selectedValue: 'wantToRead',
            ...routeProps
          }
        : shelfTitle === 'Read' && {
            shelfTitle,
            shelf: this.state.read,
            changeBookshelf: this.changeBookshelf,
            selectedValue: 'read',
            ...routeProps
          };

  returnCurrentBookCollectionViewProps = routeProps => ({
    returnMyReadsProps: this.returnMyReadsProps,
    ...routeProps
  });

  render() {
    return (
      <Router>
        <div className="app">
          <Switch>
            <Route
              path="/my-reads/search"
              render={routeProps => (
                <SearchView {...this.returnSearchViewProps(routeProps)} />
              )}
            />
            <Route
              exact
              path="/my-reads"
              render={routeProps => (
                <CurrentBookCollectionView
                  {...this.returnCurrentBookCollectionViewProps(routeProps)}
                />
              )}
            />
            <Route component={NoMatch} />
          </Switch>
        </div>
      </Router>
    );
  }
}
