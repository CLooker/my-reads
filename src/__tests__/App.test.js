import React from 'react';
import { mount, shallow } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import * as BooksAPI from '../components/BooksAPI';
import App from '../components/App';

describe('App', () => {
  const books = [{ Author: 'Chad Looker' }];
  const response = JSON.stringify({ books });
  beforeEach(() => fetch.resetMocks());

  describe('`componentDidMount`', () => {
    it('should call the correct fn', () => {
      const spy = jest.spyOn(App.prototype, 'getShelfAndRender');
      fetch.once(response);
      const mountedComponent = mount(
        <MemoryRouter>
          <App />
        </MemoryRouter>
      );
      expect(spy).toHaveBeenCalled();
    });
  });

  let component;

  beforeEach(() => {
    component = shallow(<App />);
  });

  describe('initialization', () => {
    it('should render correctly', () => {
      expect(component).toMatchSnapshot();
    });

    it('should have a particular state', () => {
      const actual = component.state();
      const expected = {
        shelf: [],
        shelfKeyValueStore: {},
        currentlyReading: [],
        wantToRead: [],
        read: []
      };
      expect(actual).toEqual(expected);
    });

    describe('`componentDidMount`', () => {
      it('should call the correct fn', () => {
        const spy = jest.spyOn(App.prototype, 'getShelfAndRender');
        fetch.once(response);
        const mountedComponent = mount(
          <MemoryRouter>
            <App />
          </MemoryRouter>
        );
        expect(spy).toHaveBeenCalled();
      });
    });
  });
});
