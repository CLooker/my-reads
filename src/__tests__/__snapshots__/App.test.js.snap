// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App initialization should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <div
        className="app"
      >
        <Switch>
          <Route
            path="/my-reads/search"
            render={[Function]}
          />
          <Route
            exact={true}
            path="/my-reads"
            render={[Function]}
          />
          <Route
            component={[Function]}
          />
        </Switch>
      </div>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Switch>
          <Route
            path="/my-reads/search"
            render={[Function]}
          />
          <Route
            exact={true}
            path="/my-reads"
            render={[Function]}
          />
          <Route
            component={[Function]}
          />
        </Switch>,
        "className": "app",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Route
              path="/my-reads/search"
              render={[Function]}
            />,
            <Route
              exact={true}
              path="/my-reads"
              render={[Function]}
            />,
            <Route
              component={[Function]}
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "path": "/my-reads/search",
              "render": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "exact": true,
              "path": "/my-reads",
              "render": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "component": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": "div",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <div
          className="app"
        >
          <Switch>
            <Route
              path="/my-reads/search"
              render={[Function]}
            />
            <Route
              exact={true}
              path="/my-reads"
              render={[Function]}
            />
            <Route
              component={[Function]}
            />
          </Switch>
        </div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Switch>
            <Route
              path="/my-reads/search"
              render={[Function]}
            />
            <Route
              exact={true}
              path="/my-reads"
              render={[Function]}
            />
            <Route
              component={[Function]}
            />
          </Switch>,
          "className": "app",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Route
                path="/my-reads/search"
                render={[Function]}
              />,
              <Route
                exact={true}
                path="/my-reads"
                render={[Function]}
              />,
              <Route
                component={[Function]}
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "path": "/my-reads/search",
                "render": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "exact": true,
                "path": "/my-reads",
                "render": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "component": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        "type": "div",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
